pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/store/SingleFileIndexDirectory.java

```diff
@@ -327,51 +336,176 @@ private void persistIndexMap(IndexEntry entry)
       throws IOException {
     File mapFile = new File(_segmentDirectory, V1Constants.INDEX_MAP_FILE_NAME);
     try (PrintWriter writer = new PrintWriter(new BufferedWriter(new FileWriter(mapFile, true)))) {
-      String startKey = getKey(entry.key.name, entry.key.type.getIndexName(), true);
-
-      StringBuilder sb = new StringBuilder();
-      sb.append(startKey).append(" = ").append(entry.startOffset);
-      writer.println(sb.toString());
-
-      String endKey = getKey(entry.key.name, entry.key.type.getIndexName(), false);
-      sb = new StringBuilder();
-      sb.append(endKey).append(" = ").append(entry.size);
-      writer.println(sb.toString());
+      persistIndexMap(entry, writer);
     }
   }
 
-  private String getKey(String column, String indexName, boolean isStartOffset) {
-    return column + MAP_KEY_SEPARATOR + indexName + MAP_KEY_SEPARATOR + (isStartOffset ? "startOffset" : "size");
-  }
-
   private String allocationContext(IndexKey key) {
     return this.getClass().getSimpleName() + key.toString();
   }
 
+  /**
+   * This method sweeps the indices marked for removal. Exception is simply bubbled up w/o
+   * trying to recover disk states from failure. This method is expected to run during segment
+   * reloading, which has failure handling by creating a backup folder before conduct reloading.
+   */
+  private void cleanupRemovedIndices()
+      throws IOException {
+    if (!_shouldCleanupRemovedIndices) {
+      return;
+    }
+
+    // To keep track of indices to be retained and put them together
+    // compactly in the new index file.
+    long nextOffset = 0;
+    List<IndexEntry> retained = new ArrayList<>();
+    File tmpIdxFile = new File(_segmentDirectory, V1Constants.INDEX_FILE_NAME + ".tmp");
+
+    // With FileChannel, we can seek to the data flexibly.
+    try (FileChannel srcCh = new RandomAccessFile(_indexFile, "r").getChannel();
```

Comment:
its better to iterate over the index entries and simply copy existing ones and skipping non-existing ones. This will make this method generic and also future proof as we add more indices.

Reply:
yeah, the changes here are pretty much doing what you've mentioned. The two copy methods below (L368, L371) iterate through the indices still in [_columnEntries](https://github.com/apache/pinot/blob/master/pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/store/SingleFileIndexDirectory.java#L82) (which keep track of indices should be `retained`), just that it tries to store forward indices firstly, followed by all kinds of inverted indices, in the file, following the current convention of day layout.

Reply:
I was thinking more about using the ColumnIndexType enum and iterating over all the columns. In other words, we should try to remove referring to DICTIONARY/FORWARD INDEX in this class. I see that we are referring to LUCENE before your change and that's already breaking the abstraction. cc @siddharthteotia we should try to remove the LUCENE reference from this class.

Reply:
One side effect to loop over all column names and all index types in ColumnIndexType enum is that the data layout of index file will change to: either key'ed by column name or key'ed by index type (depends on which is the outer loop). But today the data layout is all columns' dictionary+forward data followed by other index data, like below (which I think is determined by the logic in the [SegmentV1V2ToV3FormatConverter](https://github.com/apache/pinot/blob/master/pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/converter/SegmentV1V2ToV3FormatConverter.java#L150), and it explicitly refers to index type for the layout)
```
[col1.dictionary]
[col1.forward index]
[col2.dictionary]
[col2.forward index]
[col3.forward index] // e.g. non-dictionary encoded
--------------------------
[col1.inverted index]
[col2.json index]
[col1.bloom filter]
...
```

I don't find there is logic assumes the data layout has to like the above, but was trying to keep it just in case. If there is no risk about changing data layout, I can remove the direct reference to specific index type.

P.S. the code in [SegmentV1V2ToV3FormatConverter](https://github.com/apache/pinot/blob/master/pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/converter/SegmentV1V2ToV3FormatConverter.java#L150) seems having a bug as it doesn't copy FST, H3 and BloomFilter index during conversion. This shouldn't affect correctness, as Pinot server would re-generate them upon reloading the segment, just that it incurs unnecessary overhead.

Reply:
That's ok. Order is not really important. If we really want to, we can try to make the order deterministic by doing two passes over the old index entries and new index entries.